From: Chad Monroe <chad@monroe.io>
Date: Sun, 11 Jan 2026 13:28:01 -0800
Subject: [PATCH] pcie-mediatek-gen3: add WiFi reset GPIO deassert delay

Add support for a configurable delay after deasserting the WiFi chip
reset via the wifi-deassert-msleep DT property. Some WiFi chips require
additional stabilization time after reset release before the PCIe link
can be established.

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 drivers/pci/controller/pcie-mediatek-gen3.c |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

--- a/drivers/pci/controller/pcie-mediatek-gen3.c
+++ b/drivers/pci/controller/pcie-mediatek-gen3.c
@@ -165,6 +165,7 @@ struct mtk_msi_set {
  * @num_clks: PCIe clocks count for this port
  * @wifi_reset: reset pin for WIFI chip
  * @wifi_reset_delay_ms: delay time for WIFI chip reset
+ * @wifi_deassert_delay_ms: delay time after WIFI chip reset
  * @irq: PCIe controller interrupt number
  * @saved_irq_state: IRQ enable state saved at suspend time
  * @irq_lock: lock protecting IRQ register access
@@ -188,6 +189,7 @@ struct mtk_gen3_pcie {
 
 	struct gpio_desc *wifi_reset;
 	u32 wifi_reset_delay_ms;
+	u32 wifi_deassert_delay_ms;
 
 	int irq;
 	u32 saved_irq_state;
@@ -410,10 +412,11 @@ static int mtk_pcie_startup_port(struct
 	val |= PCIE_DISABLE_DVFSRC_VLT_REQ;
 	writel_relaxed(val, pcie->base + PCIE_MISC_CTRL_REG);
 
-	if (pcie->wifi_reset) {
+	if (!IS_ERR_OR_NULL(pcie->wifi_reset)) {
 		gpiod_set_value_cansleep(pcie->wifi_reset, 1);
 		msleep(pcie->wifi_reset_delay_ms);
 		gpiod_set_value_cansleep(pcie->wifi_reset, 0);
+		msleep(pcie->wifi_deassert_delay_ms);
 	}
 
 	/* Assert all reset signals */
@@ -892,6 +895,12 @@ static int mtk_pcie_parse_port(struct mt
 		}
 	}
 
+	ret = of_property_read_u32(dev->of_node, "wifi-deassert-msleep",
+				   &pcie->wifi_deassert_delay_ms);
+
+	if (ret)
+		pcie->wifi_deassert_delay_ms = 0;
+
 	return 0;
 }
 
